xapian-core/tests/api_weight.cc

```diff
@@ -1392,6 +1395,7 @@ class CheckStatsWeight : public Xapian::Weight {
 	TEST_REL(uniqueterms,>=,1);
 	TEST_REL(uniqueterms,<=,doclen);
 	TEST_REL(wdf,<=,wdf_upper);
+	TEST_REL(wdfdocmax,>=,1);
```

Comment:
We can definitely check more than this (and I'm not sure this condition is actually correct - if all the wdf values are 0 in a document then wdfdocmax would be 0).

For example, wdfdocmax should be the largest wdf in the document, so wdfdocmax => wdf.

But it should be equal to some wdf value, so wdfdocmax <= wdf_upper.

There are probably some others.

Reply:
I add `TEST_REL(wdfdocmax,>=,1)` assuming `TEST_REL(uniqueterms,>=,1)` is correct.
I think it would make sense to remove both of these.

wdfdocmax <= wdf_upper is not true in general. (Suppose the current term has Coll freq 1, then wdf_upper takes value 2. But wdfdocmax can take a value greater than 2 (due to some other term in the same doc.))

For wdfdocmax => wdf, wdfdocmax has to be made synonym aware.(I am figuring how to do that.)

I have added `TEST_REL(wdfdocmax,<=doclen)` for now.

Reply:
`TEST_REL(uniqueterms,>=,1);` is indeed interesting.  The case of a document not indexed by any terms isn't relevant here, as no terms means `get_sumpart()` won't get called, so the question comes down to whether uniqueterms is just the number of different terms in the document (in which case it counts terms which have zero wdf too, and the check is OK) or whether it only counts terms with non-zero wdf.  I thought we'd actually settled on the latter interpretation (since it seems undesirable that adding more filter terms changes the relevance weights, and it's also sometimes handy that the document length provides an upper bound on the number of unique terms), but the API documentation currently says:

> This is the number of different terms which index the given document.

If this count includes zero wdf terms too, then `TEST_REL(uniqueterms,<=,doclen);` is actually wrong instead.

I'll dig into this a bit more, but I think we probably just failed to clarify the API docs when we decided on this interpretation so let's assume that here for now.

Rather than removing these two `>= 1` checks, I think we should conditionalise them on `if (doclen > 0)` since `doclen` is the sum of all the wdfs in the document, so if that's non-zero there must be at least one non-zero wdf value in the document.

Reply:
I have now conditionalise it.
Also added `wdfdocmax>=wdf`  by setting `wdfdocmax` to `doclen` in case of synonym.